### ISSUE TITLE
Issues with Anonymous auth

### ISSUE THREAD

**Issue by dmtrjsg** [Joystream/atlas#4766]
- #4764
- #4765
- 4651
- #4723

**Comment by WRadoslaw** [comment]
Currently, we have a lot of issues that have their source in the fact that Orion blocks any requests that do not have an active session cookie on them (a few examples in the description). 

To my knowledge, anonymous auth is a feature that allows Orion to collect Atlas actions (video view, like, etc.) of a user who is not logged in, and later on sign up reassign them to an account, and utilize them in the process of content recommendation. Unfortunately, Atlas was not designed to work with such a concept and the process of adaptation is tough, as issues were hard to predict at the point of designing the feature and it's not much easier even now.

What if we were to enable unauthorized access to GraphQL the same as before the Orion accounts. All requests for the content of the viewer side of the application should be accessible without having a session cookie in our browser, and on top of that add a session cookie layer. 

At the page entry, we would simultaneously fire a query for the current auth user (GetAccountData) and any content (depending on the page). Even if the user is a first-timer and the first one fails, the second will return the content. Atlas will notice that the first query failed, so based on this information it will check whether there is `anonymousUserId` in local storage if it is, it will attach it to an anonymous-auth request, Orion will find this user, prologs his session and return current session cookie. Alternatively, if `anonymousUserId` wouldn't be found, Orion will receive empty params, create a new anonymous user entry, and return userId (it will be saved to local storage) together with a session cookie.

![image](https://github.com/Joystream/atlas/assets/92513933/d4321cc6-78a4-47f3-9b38-300e847b8af1)

The only problem is that I'm not sure whether there is some crucial functionality that will be killed by this approach as I didn't implement the Orion part of the accounts logic.

**Comment by bedeho** [comment]
Thank you for this, I'm just asking some initial question to get familiar here.

> Currently, we have a lot of issues that have their source in the fact that Orion blocks any requests that do not have an active session cookie on them

I did not actually see what the problems were? Are you referring to the issues above? if so, they all seem closed, so is there any lingering issue?

> Orion blocks any requests that do not have an active session cookie on them

What requests are we talking about here, is it all calls? and what do you mean by blocks? does it mean that it returns some sort of empty or error result?

What do you mean by _blocks_? Does it mean that the request literally returns null or error result?

Is there any reason Atlas is not wise enough to not bother sending this, and instead makes sure to do auth steps first? alternatively, that Orion can do auto auth for user if there is no valid anon session?

> Unfortunately, Atlas was not designed to work with such a concept and the process of adaptation is tough, as issues were hard to predict at the point of designing the feature and it's not much easier even now.

I did not understand what was meant by this.

> What if we were to enable unauthorized access to GraphQL the same as before the Orion accounts. All requests for the content of the viewer side of the application should be accessible without having a session cookie in our browser, and on top of that add a session cookie layer.

I did not grok what alternative you are foregoing by picking this approach? what is it that is worse than this which either currently is in place, or was intended to be done?

**Comment by dmtrjsg** [comment]
@bedeho 
> they all seem closed
The problems are still there, I have closed them all to open this overarching issue that encapsulates all of the cases covered in the linked issues.

**Comment by WRadoslaw** [comment]
After a long discussion, we agreed that there are 2 viable approaches.

1. Introduce extra endpoint that will serve as a point of truth when determining whether user is authorized. 

- Block all GraphQL queries
- Call Orion auth `/me`, it will return either user data or 401
- If prev step returned 401, call Orion anonymous auth to get the user authorized
- Unblock GraphQL calls 

Cons: 
- there will need to be some kind of loading state when the user is being authenticated (which will take 200-400ms on a quick connection)


2. Go with the solution proposed in my first comment and handle future distribution auth logic with something else than the session cookie.

Pros: 
- no changes on the Atlas side 
- Atlas doesn't have to wait for any auth calls, so queries for page content can be executed instantly

Cons: 
- more changes on Orion

**Comment by bedeho** [comment]
> Introduce extra endpoint that will serve as a point of truth when determining whether user is authorized.

I don't understand why we, uniquely, would need to incur 200-400ms loading latency on our app when it is very standard across all sorts of webapps to make sure users are authenticated before showing content? If I want to view my Twitter or FB feed in the browser, this is instant, and the backend will have to make sure it knows who I am, and what data I'm should see and what not. This seems totally standard, and we should not really have to invent anything new or suffer UX penalizing problems

**What is it about our situation that prevents us from just following the industry standard?**

I think its perfectly fine for us to pick any arbitrary short-cut as a fast way to solve a problem, in particular if we have other urgent tasks, but not sure that is what is motivating the current proposed alternative.

**Comment by bedeho** [comment]
I think I'd like to at least see a proposal for how we actually will do authentication of ephemeral accounts, so that we can reach our goals of having server-side state assocaited with recommendations and infrastructure access, in the future, even if we don't do it now, and that such a proposal clearly does not incur some arbitrary latency penalty, as this really should not be needed.

If have that, then any other more expedient solution which does not put us on some really conflicting codepath should be fine in my opinion.

**Comment by dmtrjsg** [comment]
There's also this issue
- https://github.com/Joystream/atlas/issues/4417

**Comment by dmtrjsg** [comment]
Agreed to move forward as per the original suggestion plus keeping the session id in the browser to count the views.
